### PR TITLE
TorchTitan e2e test on torchcomms device mesh

### PR DESCRIPTION
**Pull request by mori360**
Summary:
Composability testing with TorchComms and distributed training in TorchTitan.
  - Training with `torchcomms.new_comm`
  - Device mesh initialization with `torchcomms.init_device_mesh`
  - Integration and testing with `fully_shard`

Differential Revision: D82171763

Test plan:
TEST_BACKEND=nccl TRAIN_FILE=torchtitan.experiments.torchcomms.train  ./run_train.sh --model.name torchcomms

Loss curve:
running 1000 steps on llama3_8b.toml

Screenshot 2025-10-13 at 4 14 46 PM

### CODE REVIEW

**Comment by mori360** [comment]
> Thanks for the PR! The change looks interesting!
> 
> Is this PR for exploration or is it ready to ship to the community? If it's the former, could be start with branch / fork, instead of experiments?

Sorry for the late reply, it's planed to ship to the community as a use case for torchcomms.

> simplify the code, as I believe a lot existing components could be reused

I was cleaning the code, the main change here is
1. init communication and device mesh in (class ParallelDimsForComms in) parallel_dims.py
2. call ParallelDimsForComms in train.py
For train.py, all the other parts are reused except __init__

> set up PoC for this folder

How can I set up the PoC?

**Comment by tianyu-l** [comment]
> How can I set up the PoC?

I'll do this with a PR shortly. Should we assign you as the PoC?

**Comment by mori360** [comment]
> > How can I set up the PoC?
> 
> I'll do this with a PR shortly. Should we assign you as the PoC?

Yeah, please. There would be some further changes to enable other parallelisms and relative tests.

File: torchtitan/experiments/torchcomms/README.md [hunk 1]
```diff
@@ -0,0 +1,20 @@
+# TorchTitan & TorchComms Composability Testing
+
+This repository provides a framework for composability testing with **TorchComms** and distributed training in **TorchTitan**. The goal is to enable flexible experimentation with distributed communication primitives and parallelism strategies in PyTorch.
+---
+#### Example
+```bash
+TEST_BACKEND=nccl ./run_train.sh --model.name torchcomms
```
Comment (fegin) [inline torchtitan/experiments/torchcomms/README.md]: This doesn't seem to be correct. You will at least need to specify CONFIG_FILE.

File: torchtitan/experiments/torchcomms/README.md [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+```
+---
+## Available Features
+- **Distributed Training Utilities**
+  - Training with `torchcomms.new_comm`
+  - Device mesh initialization with `torchcomms.init_device_mesh`
+- **Composability Testing**
+  - Integration and testing with `fully_shard` (FSDP)
```
Comment (fegin) [inline torchtitan/experiments/torchcomms/README.md:20]: Is this FSDP2 only? I thought you also verified it with TP. cc., @fduwjj

File: torchtitan/experiments/torchcomms/train.py [hunk 3]
```diff
@@ -0,0 +1,364 @@
...
+        self.parallel_dims = parallel_dims = ParallelDimsForComms(
+            dp_shard=parallelism_config.data_parallel_shard_degree,
+            dp_replicate=parallelism_config.data_parallel_replicate_degree,
+            cp=parallelism_config.context_parallel_degree,
+            tp=parallelism_config.tensor_parallel_degree,
+            pp=parallelism_config.pipeline_parallel_degree,
+            ep=parallelism_config.expert_parallel_degree,
+            etp=parallelism_config.expert_tensor_parallel_degree,
+            world_size=world_size,
+        )
```
Comment (fegin) [inline torchtitan/experiments/torchcomms/train.py]: iiuc, only this part of the initialization is changed. Is this correct? Or can you point out some other things you changed?

File: torchtitan/experiments/torchcomms/README.md (hunk 2 above)
Comment (mori360) [reply torchtitan/experiments/torchcomms/README.md:20]: We are working on ND now, will update readme later

File: torchtitan/experiments/torchcomms/train.py (hunk 3 above)
Comment (mori360) [reply torchtitan/experiments/torchcomms/train.py]: Yeah, had some other changes before, but now that's the only changes. 
Will try some way to call `ParallelDimsForComms` here but avoiding copy train.__init__

Comment (fegin) [reply torchtitan/experiments/torchcomms/train.py]: You can let the original Trainer have one class variable called `parallel_dims_cls` and use that variable in the __init__ to construct `self.parallel_dims = parallel_dims`. Then you can just create a CommTrainer and replace that class variable.

Another approach is to make the following code as a method, `def create_parallel_dims(self, config) -> None:`.

```
        self.parallel_dims = parallel_dims = ParallelDimsForComms(
            dp_shard=parallelism_config.data_parallel_shard_degree,
            dp_replicate=parallelism_config.data_parallel_replicate_degree,
            cp=parallelism_config.context_parallel_degree,
            tp=parallelism_config.tensor_parallel_degree,
            pp=parallelism_config.pipeline_parallel_degree,
            ep=parallelism_config.expert_parallel_degree,
            etp=parallelism_config.expert_tensor_parallel_degree,
            world_size=world_size,
        )
```

Both approaches should work. cc., @tianyu-l @wwwjn

Comment (tianyu-l) [reply torchtitan/experiments/torchcomms/train.py]: Sounds OK. I prefer the second option as it sounds a bit more straightforward. Maybe should call it `_create_parallel_dims` as it's not supposed to be called outside.

File: torchtitan/experiments/torchcomms/README.md (hunk 2 above)
Comment (fduwjj) [reply torchtitan/experiments/torchcomms/README.md:20]: @fegin there are still some gaps on the N-D side, so we aim at first merging this PR with 1D only. This is to scale down the scope of this PR and then we will have more PRs down the road.

**Comment by mori360** [comment]
> @mori360 Once #1859 lands, please rebase and add owner to the experiment

Thanks for the reminder, can I add @d4l3k and @fduwjj  as the owner as well?

File: torchtitan/experiments/torchcomms/integration_test.py [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+        args.config_path
+    ), f"Base config path {args.config_path} does not exist"
+
+    for test_flavor in test_list:
+        # Filter by test_name if specified
+        if args.test_name != "all" and test_flavor.test_name != args.test_name:
+            continue
+
+        # Check if we have enough GPUs
+        print("run single test")
```
Comment (fduwjj) [inline torchtitan/experiments/torchcomms/integration_test.py]: is this expected?

**Comment by fduwjj** [comment]
looks like the linter is still failing? Also since you already imported into fbcode. So you want to import into fbcode as well.

File: torchtitan/experiments/README.md [hunk 5]
```diff
@@ -27,3 +27,4 @@ We provide this `experiments/` folder to host experiments that add significant v
 | [simple_fsdp](./simple_fsdp/) | [![SimpleFSDP 8 GPU Integration Tests](https://github.com/pytorch/torchtitan/actions/workflows/integration_test_8gpu_simple_fsdp.yaml/badge.svg?branch=main)](https://github.com/pytorch/torchtitan/actions/workflows/integration_test_8gpu_simple_fsdp.yaml?query=branch%3Amain) | [@ruisizhang123](https://github.com/ruisizhang123) [@tianyu-l](https://github.com/tianyu-l) |
 | [vlm](./vlm/) | [![VLM 8 GPU Integration Tests](https://github.com/pytorch/torchtitan/actions/workflows/integration_test_8gpu_vlm.yaml/badge.svg?branch=main)](https://github.com/pytorch/torchtitan/actions/workflows/integration_test_8gpu_vlm.yaml?query=branch%3Amain) | [@lkhphuc](https://github.com/lkhphuc) |
 | [forge](./forge/) | TBA | [@allenwang28](https://github.com/allenwang28) [@ebsmothers](https://github.com/ebsmothers) [@joecummings](https://github.com/joecummings) [@pbontrager](https://github.com/pbontrager) |
+| [torchcomms](./torchcomms/) | TBA | [@d4l3k](https://https://github.com/d4l3k) [@fduwjj](https://github.com/fduwjj) [@mori360 ](https://github.com/mori360) |
```
Comment (fduwjj) [inline torchtitan/experiments/README.md:30]: update with the integration test you added?

File: torchtitan/experiments/torchcomms/parallel_dims.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+            # TODO: it's a hacky solution for now and we will update it in a week
+            mesh = init_device_mesh(
+                mesh_dim_comms=(comm, comm, comm, comm),
+                mesh_dim_names=("dp_shard", "dp", "dp_cp", "dp_shard_cp"),
+                _global_comm=comm,
+            )
+            self.comms.append(comm)
+            return mesh
+        else:
+            raise NotImplementedError("Only support DP shard parallelism for now.")
```
Comment (fduwjj) [inline torchtitan/experiments/torchcomms/parallel_dims.py]: It should be "Only support FSDP 1D parallelism for now."?

File: torchtitan/experiments/torchcomms/README.md [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+---
+## To Be Added
+- Composability testing with additional parallelism strategies (e.g., tensor, pipeline, model parallelism)
```
Comment (fduwjj) [inline torchtitan/experiments/torchcomms/README.md]: Shall we also add integration with N-D parallelism and e2e test?

Comment (fduwjj) [reply torchtitan/experiments/torchcomms/README.md]: For example "Integration and testing with additional parallelism strategies (e.g., tensor, pipeline, model parallelism) other than fully_shard"?

File: torchtitan/experiments/README.md (hunk 5 above)
Comment (mori360) [reply torchtitan/experiments/README.md:30]: discussed offline, will add the integration test soon after fixing the dependency

File: torchtitan/experiments/torchcomms/requirements.txt [hunk 8]
```diff
@@ -0,0 +1 @@
+# TODO: add dependencies
```
Comment (fduwjj) [inline torchtitan/experiments/torchcomms/requirements.txt]: Nit: "add dependencies for TorchComm once the pip release is ready".

Because all other dependencies are not blocked.

File: torchtitan/experiments/torchcomms/integration_tests.py [hunk 9]
```diff
@@ -0,0 +1,125 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
```
Comment (fduwjj) [inline torchtitan/experiments/torchcomms/integration_tests.py]: As discussed offline, you don't have this one in this PR? Let's remove it completely from this PR? because currently this one is not working because of import torchcomms?

**Comment by fduwjj** [comment]
also looks like your PR is also not fully sync with the diff. You need to import it every time when you update your PR.

**Comment by fduwjj** [comment]
Also we will have more converge and perf test down the road as follow-up PRs.

File: torchtitan/experiments/torchcomms/README.md [hunk 10]
```diff
@@ -0,0 +1,20 @@
+# TorchTitan & TorchComms Composability Testing
+
+This repository provides a framework for composability testing with **TorchComms** and distributed training in **TorchTitan**. The goal is to enable flexible experimentation with distributed communication primitives and parallelism strategies in PyTorch.
```
Comment (tianyu-l) [inline torchtitan/experiments/torchcomms/README.md]: This is currently in bold font and look a bit obtrusive. Could you adjust them to use plain font?

Comment (tianyu-l) [inline torchtitan/experiments/torchcomms/README.md]:
```suggestion
This folder provides a framework for composability testing with **TorchComms** and distributed training in **TorchTitan**. The goal is to enable flexible experimentation with distributed communication primitives and parallelism strategies in PyTorch.
```

File: torchtitan/experiments/torchcomms/README.md [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+---
+#### Example
```
Comment (tianyu-l) [inline torchtitan/experiments/torchcomms/README.md:8]: mention that the command below uses Llama 3 as an example, but should work on all models.

File: torchtitan/experiments/torchcomms/README.md [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+```bash
+TEST_BACKEND={backend} TRAIN_FILE=torchtitan.experiments.torchcomms.train ./run_train.sh --model.name torchcomms
```
Comment (tianyu-l) [inline torchtitan/experiments/torchcomms/README.md]: > TEST_BACKEND={backend}

What should this be?

File: torchtitan/experiments/torchcomms/__init__.py [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+from torchtitan.components.loss import build_cross_entropy_loss
+from torchtitan.components.lr_scheduler import build_lr_schedulers
+from torchtitan.components.optimizer import build_optimizers
+from torchtitan.components.tokenizer import build_hf_tokenizer
+from torchtitan.datasets.hf_datasets import build_hf_dataloader
+from torchtitan.models.llama3 import llama3_configs, pipeline_llama, Transformer
+from torchtitan.models.llama3.infra.parallelize import parallelize_llama
+from torchtitan.protocols.train_spec import register_train_spec, TrainSpec
+
+register_train_spec(
```
Comment (tianyu-l) [inline torchtitan/experiments/torchcomms/__init__.py]: why do you need to register this `TrainSpec`?

File: torchtitan/experiments/torchcomms/llama3_8b.toml
Comment (tianyu-l) [inline torchtitan/experiments/torchcomms/llama3_8b.toml:1]: why do you need to add this toml config?

File: torchtitan/experiments/torchcomms/README.md (hunk 12 above)
Comment (tianyu-l) [inline torchtitan/experiments/torchcomms/README.md]: Let's set `CONFIG_FILE` here, too. You can refer to examples in main README.md

File: torchtitan/train.py [hunk 14]
```diff
@@ -376,6 +369,18 @@ def __init__(self, job_config: JobConfig):
             f"(warmup {job_config.lr_scheduler.warmup_steps})"
         )
 
+    def create_parallel_dims(self, parallelism_config, world_size) -> ParallelDims:
```
Comment (tianyu-l) [inline torchtitan/train.py]:
```suggestion
    def _create_parallel_dims(self, parallelism_config, world_size) -> ParallelDims:
```

File: torchtitan/experiments/torchcomms/train.py [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+from torchtitan.tools.logging import init_logger, logger
+from torchtitan.train import Trainer
+
+from .parallel_dims import ParallelDimsForComms
+
+
+class CommsTrainer(Trainer):
+    parallel_dims: ParallelDimsForComms
+
+    def create_parallel_dims(self, parallelism_config, world_size) -> ParallelDims:
```
Comment (tianyu-l) [inline torchtitan/experiments/torchcomms/train.py]:
```suggestion
    def _create_parallel_dims(self, parallelism_config, world_size) -> ParallelDims:
```

File: torchtitan/experiments/torchcomms/train.py [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+
+from torchtitan.config import ConfigManager
+from torchtitan.distributed import ParallelDims
+from torchtitan.tools.logging import init_logger, logger
+from torchtitan.train import Trainer
+
+from .parallel_dims import ParallelDimsForComms
+
+
+class CommsTrainer(Trainer):
```
Comment (tianyu-l) [inline torchtitan/experiments/torchcomms/train.py]:
```suggestion
class TorchCommsTrainer(Trainer):
```

File: torchtitan/experiments/torchcomms/parallel_dims.py [hunk 17]
```diff
@@ -0,0 +1,58 @@
...
+from torchcomms.device_mesh import init_device_mesh
+
+from torchtitan.distributed.parallel_dims import ParallelDims
+from torchtitan.tools.logging import logger
+
+__all__ = ["ParallelDimsForComms"]
+
+
+@dataclass
+class ParallelDimsForComms(ParallelDims):
```
Comment (tianyu-l) [inline torchtitan/experiments/torchcomms/parallel_dims.py]:
```suggestion
class TorchCommsParallelDims(ParallelDims):
```

File: torchtitan/experiments/torchcomms/README.md (hunk 12 above)
Comment (mori360) [reply torchtitan/experiments/torchcomms/README.md]: users can input backend they want to use, e.g. nccl or other backend
It's a bit confusing here, will change to TEST_BACKEND=nccl

Comment (mori360) [reply torchtitan/experiments/torchcomms/README.md]: users can input backend they want to use, e.g. nccl or other backend
It's a bit confusing here, will change to TEST_BACKEND=nccl

Comment (tianyu-l) [reply torchtitan/experiments/torchcomms/README.md]: Can we mention all the available backends? From the readme it's hard to tell what people should put here.

File: torchtitan/experiments/torchcomms/llama3_8b.toml
Comment (tianyu-l) [reply torchtitan/experiments/torchcomms/llama3_8b.toml:1]: shall we remove this file?

File: torchtitan/experiments/torchcomms/__init__.py
Comment (tianyu-l) [inline torchtitan/experiments/torchcomms/__init__.py:1]: I think we can remove this file for now.

File: torchtitan/experiments/torchcomms/README.md [hunk 18]
```diff
@@ -0,0 +1,22 @@
+# TorchTitan & TorchComms Composability Testing
+
+This folder provides a framework for composability testing with TorchComms and distributed training in TorchTitan. The goal is to enable flexible experimentation with distributed communication primitives and parallelism strategies in PyTorch.
```
Comment (tianyu-l) [inline torchtitan/experiments/torchcomms/README.md:5]: seems font hasn't been fixed

File: torchcomms
Comment (tianyu-l) [inline torchcomms:1]: what's this module?

File: torchtitan/experiments/torchcomms/README.md (hunk 12 above)
Comment (fduwjj) [reply torchtitan/experiments/torchcomms/README.md]: let's mention nccl, gloo or any other user defined customized backend for now. Also let's mention that the user customized backend needs to implement torchComm wrapper. (We just don't mention the backend which cannot be mentioned at this moment.)

File: torchcomms
Comment (fduwjj) [reply torchcomms:1]: For now, we cannot mention too much details. We will add more context when it goes public. We need to merge this PR first so that the titan integration can go with the release of torchcomm.

@mori360 let's add a TODO here to add more explanation once the torchcomm goes public.

**Comment by fduwjj** [comment]
Looks like you have lint error as well?

**Comment by fduwjj** [comment]
if you choose this:

image


Would that help make CI happy?

File: torchtitan/experiments/torchcomms/README.md [hunk 19]
```diff
@@ -0,0 +1,25 @@
...
+---
+### Available Features
+- **Distributed Training Utilities**
+  - Training with `torchcomms.new_comm`
+  - Device mesh initialization with `torchcomms.init_device_mesh`
+- **Composability Testing**
+  - Integration and testing with `fully_shard` (FSDP)
+---
+### To Be Added
+- Integration and testing with additional parallelism strategies (e.g., tensor, pipeline, model parallelism) other than fully_shard
```
Comment (fduwjj) [inline torchtitan/experiments/torchcomms/README.md]: can you remove model parallelism or replace it with context parallelism? Thanks
